File: mc/domain/workflow_contract.py
"""
Workflow contract adapter — loads the canonical workflow spec and exposes helpers.

The single source of truth is ``shared/workflow/workflow_spec.json``.
This module loads that file once at import time and provides Python-friendly
accessor functions used by the rest of the ``mc`` package.

Story 15.1 — Shared Workflow Contract.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

# ---------------------------------------------------------------------------
# Load spec once at import time
# ---------------------------------------------------------------------------

_SPEC_PATH = Path(__file__).resolve().parents[2] / "shared" / "workflow" / "workflow_spec.json"

with open(_SPEC_PATH, "r") as _f:
    SPEC: dict[str, Any] = json.load(_f)

# ---------------------------------------------------------------------------
# Derived constants
# ---------------------------------------------------------------------------

TASK_STATUSES: list[str] = SPEC["taskStatuses"]
STEP_STATUSES: list[str] = SPEC["stepStatuses"]

_TASK_TRANSITIONS: dict[str, list[str]] = SPEC["taskTransitions"]
_TASK_UNIVERSAL_TARGETS: list[str] = SPEC["taskUniversalTargets"]
_TASK_TRANSITION_EVENTS: dict[str, str] = SPEC["taskTransitionEvents"]
_TASK_UNIVERSAL_TARGET_EVENTS: dict[str, str] = SPEC["taskUniversalTargetEvents"]

_STEP_TRANSITIONS: dict[str, list[str]] = SPEC["stepTransitions"]
_STEP_TRANSITION_EVENTS: dict[str, str] = SPEC["stepTransitionEvents"]

_MENTION_SAFE_STATUSES: set[str] = set(SPEC["mentionSafeTaskStatuses"])

THREAD_MESSAGE_TYPES: list[str] = SPEC["threadMessageTypes"]


# ---------------------------------------------------------------------------
# Task helpers
# ---------------------------------------------------------------------------


def is_valid_task_transition(from_status: str, to_status: str) -> bool:
    """Check if a task state transition is valid.

    Returns True if *to_status* is reachable from *from_status* either via
    the explicit transition map or via universal target statuses.
    """
    if to_status in _TASK_UNIVERSAL_TARGETS:
        return True
    allowed = _TASK_TRANSITIONS.get(from_status, [])
    return to_status in allowed


def get_allowed_transitions(status: str) -> list[str]:
    """Return the list of statuses reachable from *status* (excluding universal targets)."""
    return list(_TASK_TRANSITIONS.get(status, []))


def get_universal_transitions() -> list[str]:
    """Return the list of universal target statuses (reachable from any status)."""
    return list(_TASK_UNIVERSAL_TARGETS)


def get_task_transition_event(from_status: str, to_status: str) -> str | None:
    """Return the activity event type for a task transition, or None if unmapped.

    For universal targets, returns the universal target event.
    """
    if to_status in _TASK_UNIVERSAL_TARGET_EVENTS:
        return _TASK_UNIVERSAL_TARGET_EVENTS[to_status]
    return _TASK_TRANSITION_EVENTS.get(f"{from_status}->{to_status}")


# ---------------------------------------------------------------------------
# Step helpers
# ---------------------------------------------------------------------------


def is_valid_step_transition(from_status: str, to_status: str) -> bool:
    """Check if a step state transition is valid."""
    allowed = _STEP_TRANSITIONS.get(from_status, [])
    return to_status in allowed


def get_step_allowed_transitions(status: str) -> list[str]:
    """Return the list of step statuses reachable from *status*."""
    return list(_STEP_TRANSITIONS.get(status, []))


def get_step_transition_event(from_status: str, to_status: str) -> str | None:
    """Return the activity event type for a step transition, or None if unmapped."""
    return _STEP_TRANSITION_EVENTS.get(f"{from_status}->{to_status}")


# ---------------------------------------------------------------------------
# Mention safety
# ---------------------------------------------------------------------------


def is_mention_safe(status: str) -> bool:
    """Return True if the given task status allows @mention interactions."""
    return status in _MENTION_SAFE_STATUSES

File: dashboard/convex/lib/workflowContract.ts
/**
 * Workflow contract adapter — loads the canonical workflow spec and exposes helpers.
 *
 * The single source of truth is `shared/workflow/workflow_spec.json`.
 * This module loads that file and provides TypeScript-friendly accessor
 * functions used by Convex mutations and queries.
 *
 * Story 15.1 — Shared Workflow Contract.
 */

import spec from "../../../shared/workflow/workflow_spec.json";

// ---------------------------------------------------------------------------
// Exported types
// ---------------------------------------------------------------------------

export type TaskStatus = (typeof spec.taskStatuses)[number];
export type StepStatus = (typeof spec.stepStatuses)[number];

// ---------------------------------------------------------------------------
// Exported constants
// ---------------------------------------------------------------------------

export const TASK_STATUSES: readonly string[] = spec.taskStatuses;
export const STEP_STATUSES: readonly string[] = spec.stepStatuses;
export const THREAD_MESSAGE_TYPES: readonly string[] = spec.threadMessageTypes;

// ---------------------------------------------------------------------------
// Task helpers
// ---------------------------------------------------------------------------

/**
 * Check if a task state transition is valid.
 *
 * Returns true if `toStatus` is reachable from `fromStatus` either via
 * the explicit transition map or via universal target statuses.
 */
export function isValidTransition(
  fromStatus: string,
  toStatus: string,
): boolean {
  if (
    (spec.taskUniversalTargets as readonly string[]).includes(toStatus)
  ) {
    return true;
  }
  const allowed =
    (spec.taskTransitions as Record<string, string[]>)[fromStatus] ?? [];
  return allowed.includes(toStatus);
}

/**
 * Return the list of statuses reachable from `status` (excluding universal targets).
 */
export function getAllowedTransitions(status: string): string[] {
  return [
    ...((spec.taskTransitions as Record<string, string[]>)[status] ?? []),
  ];
}

/**
 * Return the list of universal target statuses (reachable from any status).
 */
export function getUniversalTransitions(): string[] {
  return [...spec.taskUniversalTargets];
}

/**
 * Return the activity event type for a task transition, or undefined if unmapped.
 *
 * For universal targets, returns the universal target event.
 */
export function getTaskTransitionEvent(
  fromStatus: string,
  toStatus: string,
): string | undefined {
  const universalEvent =
    (spec.taskUniversalTargetEvents as Record<string, string>)[toStatus];
  if (universalEvent) {
    return universalEvent;
  }
  const key = `${fromStatus}->${toStatus}`;
  return (spec.taskTransitionEvents as Record<string, string>)[key];
}

// ---------------------------------------------------------------------------
// Step helpers
// ---------------------------------------------------------------------------

/**
 * Check if a step state transition is valid.
 */
export function isValidStepTransition(
  fromStatus: string,
  toStatus: string,
): boolean {
  const allowed =
    (spec.stepTransitions as Record<string, string[]>)[fromStatus] ?? [];
  return allowed.includes(toStatus);
}

/**
 * Return the list of step statuses reachable from `status`.
 */
export function getStepAllowedTransitions(status: string): string[] {
  return [
    ...((spec.stepTransitions as Record<string, string[]>)[status] ?? []),
  ];
}

/**
 * Return the activity event type for a step transition, or undefined if unmapped.
 */
export function getStepTransitionEvent(
  fromStatus: string,
  toStatus: string,
): string | undefined {
  const key = `${fromStatus}->${toStatus}`;
  return (spec.stepTransitionEvents as Record<string, string>)[key];
}

// ---------------------------------------------------------------------------
// Mention safety
// ---------------------------------------------------------------------------

/**
 * Return true if the given task status allows @mention interactions.
 */
export function isMentionSafe(status: string): boolean {
  return (spec.mentionSafeTaskStatuses as readonly string[]).includes(status);
}
